File: grading_costs_live.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
from __future__ import annotations
from datetime import datetime, timezone
from copy import deepcopy
from html import unescape
from pathlib import Path
import json,re,time,urllib.request

from safe_runtime import atomic_write_json, safe_read_text, safe_urlopen

ROOT=Path(__file__).resolve().parent
CACHE=ROOT/'grading_costs_cache.json'
WATCH=ROOT/'grading_company_updates.json'
CACHE_TTL=6*60*60
UA='Mozilla/5.0 TCG-Grader/1.1'
ALLOWED_HOSTS={
 'www.psacard.com','psacard.com','www.beckett.com','beckett.com',
 'www.cgccards.com','cgccards.com','taggrading.com','www.taggrading.com',
 'break.co.kr','www.break.co.kr',
}
PREFERRED_MARKET={'PSA':'US','BGS':'US','CGC':'US','TAG':'US','BRG':'KR'}

# Last-known official baselines are fail-safe fallbacks only. The scheduled official
# watcher can overlay verified changes without mutating source code.
COMPANIES={
 'PSA':{
  'source':'https://www.psacard.com/services/tradingcardgrading','currency':'USD','shipping':'checkout_calculated','insurance':'tier_max_insured_value',
  'services':[{'name':'Regular','fee':79.99,'max_insured_value':1500},{'name':'Express','fee':149.00,'max_insured_value':2500},{'name':'Super Express','fee':349.00,'max_insured_value':5000},{'name':'Walk-Through','fee':599.00,'max_insured_value':10000},{'name':'Premium +','fee':999.00,'max_insured_value':25000}],
 },
 'BGS':{
  'source':'https://www.beckett.com/grading','currency':'USD','shipping':'checkout_calculated','insurance':'return_shipping_and_insurance_variable',
  'services':[{'name':'Base','fee':14.95,'availability':'paused'},{'name':'Base + Subgrades','fee':17.95,'availability':'paused'},{'name':'Standard','fee':34.95,'availability':'paused'},{'name':'Express','fee':79.95,'availability':'open'},{'name':'Priority','fee':124.95,'availability':'open'}],
 },
 'CGC':{
  'source':'https://www.cgccards.com/submit/services-fees/cgc-grading/?view=cards','currency':'USD','shipping':'checkout_calculated','insurance':'declared_value_tier',
  'services':[{'name':'Bulk','fee':17.00,'min_cards':25,'max_value':500},{'name':'Economy','fee':20.00,'max_value':1000},{'name':'Standard','fee':55.00,'max_value':3000},{'name':'Express','fee':100.00,'max_value':10000},{'name':'WalkThrough','fee':300.00,'max_value':100000}],
 },
 'TAG':{
  'source':'https://taggrading.com/pages/pricing','currency':'USD','shipping':'return_shipping_flat_at_checkout','insurance':'included_by_tier',
  'services':[{'name':'Basic','fee':22.00,'min_cards':10,'insurance_per_card':300},{'name':'Standard','fee':39.00,'insurance_per_card':500},{'name':'Express','fee':59.00,'insurance_per_card':1000},{'name':'Priority','fee':149.00,'insurance_per_card':2500},{'name':'Walkthrough','fee':299.00,'insurance_per_card':5000}],
  'extras':[{'name':'Submission Kit','fee':49.95,'currency':'USD','coverage_order':1000,'region':'US only'},{'name':'Return Shipping Insurance','fee':14.99,'currency':'USD','coverage_order':1000}],
 },
 'BRG':{
  'source':'https://break.co.kr/','currency':'KRW','shipping':'domestic_carrier_actual','insurance':'carrier_actual',
  'services':[{'name':'Regular','fee':19800},{'name':'Express','fee':39800},{'name':'Bulk','fee':13800,'min_cards':20},{'name':'Reholder','fee':9800},{'name':'BRG GEN','fee':9800}],
  'extras':[{'name':'오토 등급','fee':3000,'currency':'KRW'}],
 },
}

def _text(html:str)->str:
    html=re.sub(r'(?is)<script.*?</script>|<style.*?</style>',' ',html)
    html=re.sub(r'(?s)<[^>]+>',' ',html)
    return re.sub(r'\s+',' ',unescape(html)).strip()

def _fetch(url:str)->str:
    req=urllib.request.Request(url,headers={'User-Agent':UA,'Accept-Language':'en-US,en;q=0.8,ko;q=0.6'})
    with safe_urlopen(req,timeout=12,allowed_hosts=ALLOWED_HOSTS,max_redirects=2) as r:
        return _text(r.read(2_000_000).decode('utf-8','ignore'))

def _near_price(text,name,currency):
    # Search a short window after the service label; accept only plausible official prices.
    m=re.search(re.escape(name)+r'.{0,260}',text,re.I)
    if not m:return None
    win=m.group(0)
    if currency=='KRW':
        vals=[int(v.replace(',','')) for v in re.findall(r'([0-9]{1,3}(?:,[0-9]{3})+)\s*원',win)]
        return next((v for v in vals if 3000<=v<=5000000),None)
    if currency=='JPY':
        vals=[int(v.replace(',','')) for v in re.findall(r'[￥¥]\s*([0-9][0-9,]*)',win)]
        return next((v for v in vals if 500<=v<=20000000),None)
    vals=[float(v.replace(',','')) for v in re.findall(r'\$\s*([0-9][0-9,]*(?:\.[0-9]{1,2})?)',win)]
    return next((v for v in vals if 5<=v<=50000),None)

def _refresh_company(name,cfg):
    out=deepcopy(cfg); out['source_ok']=False; out['live_updates']=0
    try:text=_fetch(cfg['source'])
    except Exception as e:
        out['source_error']=type(e).__name__;return out
    out['source_ok']=True
    for svc in out.get('services',[]):
        p=_near_price(text,svc['name'],cfg['currency'])
        if p is not None and abs(float(p)-float(svc['fee']))/max(float(svc['fee']),1)<5:
            svc['fee']=p;svc['live_verified']=True;out['live_updates']+=1
    if name=='BGS':
        low=text.lower()
        if 'temporarily paused' in low or 'sold out' in low:
            for svc in out['services']:
                if svc['name'] in ('Base','Base + Subgrades','Standard'):svc['availability']='paused'
    return out

def _watch_signature():
    try:
        st=WATCH.stat();return f'{st.st_mtime_ns}:{st.st_size}'
    except OSError:return 'missing'

def _load_watch():
    try:
        data=json.loads(safe_read_text(WATCH,max_bytes=8_000_000))
        if not isinstance(data,dict) or not data.get('policy',{}).get('official_sources_only'):return {}
        return data
    except (OSError,ValueError,TypeError,UnicodeError,json.JSONDecodeError):return {}

def _merge_watch(companies,watch):
    if not watch:return
    watched=watch.get('companies',{}) if isinstance(watch.get('companies'),dict) else {}
    changes=watch.get('history',[]) if isinstance(watch.get('history'),list) else []
    for name,out in companies.items():
        company=watched.get(name,{}) if isinstance(watched.get(name),dict) else {}
        markets=company.get('markets',{}) if isinstance(company.get('markets'),dict) else {}
        out['regional_markets']=deepcopy(markets)
        preferred=PREFERRED_MARKET.get(name)
        market=markets.get(preferred,{}) if isinstance(markets.get(preferred),dict) else {}
        rows=market.get('services',[]) if isinstance(market.get('services'),list) else []
        by_name={str(s.get('name')):s for s in out.get('services',[]) if isinstance(s,dict) and s.get('name')}
        for observed in rows:
            if not isinstance(observed,dict) or not observed.get('verified_official_source') or not observed.get('name'):continue
            key=str(observed['name']);target=by_name.get(key)
            if target is None:
                target={'name':key};out.setdefault('services',[]).append(target);by_name[key]=target
            for field in ('fee','turnaround_business_days','max_declared_or_insured_value','availability'):
                if field in observed:target[field]=observed[field]
            target['live_verified']=True;target['watch_source']=observed.get('source')
        # Only an explicit verified removal record can retire a fallback service.
        removed={str(c.get('service')) for c in changes if isinstance(c,dict) and c.get('company')==name and c.get('type')=='service_removed' and c.get('verified_official_source')}
        for svc in out.get('services',[]):
            if str(svc.get('name')) in removed:svc['availability']='retired'
        out['watch_source_health']=deepcopy(company.get('source_health',[]))

def _load_cache():
    try:
        d=json.loads(safe_read_text(CACHE,max_bytes=4_000_000))
        if d.get('_watch_signature')!=_watch_signature():return None
        if time.time()-float(d.get('_epoch',0))<CACHE_TTL:return d
    except (OSError,ValueError,TypeError,UnicodeError,json.JSONDecodeError):pass
    return None

def _save_cache(data):
    try:atomic_write_json(CACHE,data,suffix='.grading-costs.tmp')
    except (OSError,ValueError,TypeError):pass

def get_grading_costs(force:bool=False):
    if not force:
        cached=_load_cache()
        if cached:
            cached['cache']='hit';return cached
    companies={k:_refresh_company(k,v) for k,v in COMPANIES.items()}
    watch=_load_watch();_merge_watch(companies,watch)
    now=datetime.now(timezone.utc).isoformat(timespec='seconds')
    data={
        'ok':True,'checked_at':now,'refresh_hours':6,'companies':companies,
        'watch_checked_at':watch.get('checked_at') if watch else None,
        'recent_changes':deepcopy(watch.get('recent_changes',[])) if watch else [],
        'grading_events':deepcopy(watch.get('announcements',[])) if watch else [],
        'watch_summary':deepcopy(watch.get('summary',{})) if watch else {},
        'notice':'공식 업체 페이지를 주기적으로 확인하며, 가격·서비스명·예상 납기·접수중지/재개·공식 이벤트 변경을 별도 이력으로 검증합니다. 공식 파싱 실패 시 마지막 검증값을 유지하고 커뮤니티/인스타 게시물만으로는 자동 변경하지 않습니다. 배송비와 일부 보험료는 실제 체크아웃/택배사 값을 사용합니다.',
        '_epoch':time.time(),'_watch_signature':_watch_signature(),'cache':'refresh',
    }
    _save_cache(data);return data

File: grading_costs_live.js
(()=>{
'use strict';
const $=id=>document.getElementById(id);
const esc=v=>String(v??'').replace(/[&<>"']/g,ch=>({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'}[ch]));
const usd=n=>`$${Number(n).toFixed(2)}`;
const krw=n=>`₩${Math.round(Number(n)).toLocaleString('ko-KR')}`;
const jpy=n=>`¥${Math.round(Number(n)).toLocaleString('ja-JP')}`;
const money=(n,c)=>c==='KRW'?krw(n):c==='JPY'?jpy(n):usd(n);
function shippingText(v){return ({checkout_calculated:'결제 시 자동계산',return_shipping_flat_at_checkout:'반송배송비 결제 시 계산',domestic_carrier_actual:'국내 택배 실제요금',return_shipping_and_insurance_variable:'반송배송·보험 실제요금',declared_value_tier:'신고가액/서비스별'}[v]||'실제 결제값 확인')}
function insuranceText(c){
 if(c.insurance==='included_by_tier')return '서비스별 보험 포함';
 if(c.insurance==='tier_max_insured_value')return '서비스별 최대 보험가액';
 if(c.insurance==='declared_value_tier')return '신고가액 한도 기준';
 if(c.insurance==='carrier_actual')return '택배사 보험 실제요금';
 return '보험/보상 실제 결제조건 확인';
}
function serviceText(x,currency){
 const fee=x.fee==null?'가격 확인 필요':money(x.fee,currency);
 const tat=x.turnaround_business_days?` · 약 ${x.turnaround_business_days}영업일`:'';
 const state=x.availability==='paused'?' · 접수중지':x.availability==='retired'?' · 종료':'';
 return `${esc(x.name)} ${esc(fee)}${tat}${state}`;
}
function regionalText(c){
 const markets=c.regional_markets||{};
 return Object.entries(markets).map(([market,m])=>{
  const rows=(m.services||[]).filter(x=>!['paused','retired'].includes(x.availability)).slice(0,5);
  if(!rows.length)return '';
  return `<small><b>${esc(market)}</b> ${rows.map(x=>serviceText(x,m.currency||c.currency)).join(' · ')}</small>`;
 }).filter(Boolean).join('');
}
function changeText(row){
 const type={service_added:'신규 서비스',service_changed:'서비스 변경',service_removed:'서비스 종료',official_announcement:'공식 공지',official_page_changed_unparsed:'공식 페이지 변경 감지'}[row.type]||row.type||'변경';
 const svc=row.service?` · ${row.service}`:'';
 if(row.type==='service_changed'&&row.changes){
  const bits=Object.entries(row.changes).map(([k,v])=>`${k}: ${v.before??'-'} → ${v.after??'-'}`);
  return `${type}${svc} · ${bits.join(', ')}`;
 }
 return `${type}${svc}${row.title?` · ${row.title}`:''}`;
}
function renderChanges(j){
 const target=$('gradingCostChanges');if(!target)return;
 const changes=(j.recent_changes||[]).slice(-8).reverse();
 const events=(j.grading_events||[]).slice(-6).reverse();
 if(!changes.length&&!events.length){target.innerHTML='<small>최근 공식 변경 이력이 없습니다. 감시기는 공식 업체 페이지를 주기적으로 확인합니다.</small>';return;}
 const a=changes.map(x=>`<li>${esc(changeText(x))}${x.url?` <a href="${esc(x.url)}" target="_blank" rel="noopener noreferrer">공식</a>`:''}</li>`).join('');
 const b=events.map(x=>`<li>${esc(x.company||'')} · ${esc(x.title||'')} <a href="${esc(x.url||'#')}" target="_blank" rel="noopener noreferrer">공식</a></li>`).join('');
 target.innerHTML=`${a?`<b>최근 요금·서비스 변경</b><ul>${a}</ul>`:''}${b?`<b>업체 이벤트·공지</b><ul>${b}</ul>`:''}`;
}
function renderLive(j){
 const box=$('gradingCostRows');if(!box)return;
 box.innerHTML=Object.entries(j.companies||{}).map(([name,c])=>{
  const active=(c.services||[]).filter(x=>!['paused','retired'].includes(x.availability));
  const tiers=active.slice(0,5).map(x=>serviceText(x,c.currency)).join(' · ');
  const ins=(active.find(x=>x.insurance_per_card)||active.find(x=>x.max_insured_value)||{});
  const cap=ins.insurance_per_card?` · 보험 ${usd(ins.insurance_per_card)}/장`:ins.max_insured_value?` · 최대보험 ${usd(ins.max_insured_value)}`:'';
  return `<article class="gcp-row"><div><b>${esc(name)}</b><span>${tiers||'요금 확인 필요'}</span><small>배송: ${esc(shippingText(c.shipping))} · 보험: ${esc(insuranceText(c))}${esc(cap)}</small>${regionalText(c)}</div><a href="${esc(c.source)}" target="_blank" rel="noopener noreferrer">공식 요금표</a></article>`;
 }).join('');
 renderChanges(j);
 $('gradingCostNote').textContent=(j.notice||'')+' · API 확인 '+(j.checked_at||'')+(j.watch_checked_at?` · 업체감시 ${j.watch_checked_at}`:'');
}
function renderStatic(w){
 const box=$('gradingCostRows');if(!box)return;
 box.innerHTML=Object.entries(w.companies||{}).map(([name,c])=>{
  const markets=c.markets||{};
  const body=Object.entries(markets).map(([market,m])=>`<span><b>${esc(market)}</b> ${(m.services||[]).slice(0,6).map(x=>serviceText(x,m.currency)).join(' · ')||'마지막 검증값 없음'}</span>`).join('');
  return `<article class="gcp-row"><div><b>${esc(name)}</b>${body}</div></article>`;
 }).join('')||'<span class="gcp-warn">업체 감시 스냅샷 준비 중</span>';
 renderChanges({recent_changes:w.recent_changes||[],grading_events:w.announcements||[]});
 $('gradingCostNote').textContent=`공식 업체 감시 정적 스냅샷 · ${w.checked_at||''}`;
}
function mount(){
 if($('gradingCostPanel'))return;
 const anchor=$('autoGradeMarketFlow'); if(!anchor)return;
 const s=document.createElement('section');s.id='gradingCostPanel';s.className='card grading-cost-panel';
 s.innerHTML='<div class="gcp-head"><div><h3>💳 업체별 감정비 · 배송 · 보험</h3><p>PSA/BGS/CGC/TAG/BRG 공식 출처 기준. 가격·서비스·납기·접수상태·이벤트 변경을 함께 감시합니다.</p></div><button id="gradingCostRefresh" type="button">요금 새로고침</button></div><div id="gradingCostRows" class="gcp-rows">불러오는 중…</div><div id="gradingCostChanges" class="gcp-note"></div><div id="gradingCostNote" class="gcp-note"></div>';
 anchor.insertAdjacentElement('afterend',s);$('gradingCostRefresh')?.addEventListener('click',load);
 load();
}
async function load(){
 const box=$('gradingCostRows'); if(!box)return;
 box.textContent='공식 요금·업체 변경사항 불러오는 중…';
 try{
  const r=await fetch('/api/grading-costs?t='+Date.now(),{cache:'no-store'});const j=await r.json();if(!r.ok||!j.ok)throw new Error('api');
  renderLive(j);return;
 }catch(_){}
 try{
  const r=await fetch('/grading_company_updates.json?t='+Date.now(),{cache:'no-store'});const w=await r.json();if(!r.ok||!w.policy?.official_sources_only)throw new Error('static');
  renderStatic(w);return;
 }catch(_){box.innerHTML='<span class="gcp-warn">공식 업체 감시 자료를 불러오지 못했습니다. 서버 연결 또는 다음 정적 갱신을 확인하세요.</span>'}
}
function boot(){mount();setInterval(()=>{if($('gradingCostPanel')&&document.visibilityState==='visible')load()},3*60*60*1000)}
if(document.readyState==='loading')document.addEventListener('DOMContentLoaded',boot);else boot();
})();
